File: binarytree.py
class BinaryTree(object):

    def __init__(self, data = None, left = None, right = None):
        self.data = data
        self.left = left
        self.right = right

    @property
    def data(self):
        return self.__data

    @data.setter
    def data(self, data):
        self.__data = data

    def add_leftchild(self, tree):
        if type(tree.data) == type(self.data):
            self.left = tree

    def add_rightchild(self, tree):
        if type(tree.data) == type(self.data):
            self.right = tree
        else:
            raise TypeError('Type mismatch between ' + type(self.data).__name__ + ' and ' + type(tree.data).__name__)

    def preorder(self):
        if self.data is not None:
            yield self.data
        if self.left is not None:
            for d in self.left.preorder():
                yield d
        if self.right is not None:
            for d in self.right.preorder():
                yield d

    def __iter__(self):
        output = self.preorder()
        return output

File: BinaryTree.java
import java.util.Iterator;
import java.util.LinkedList;
import java.util.NoSuchElementException;


public class BinaryTree<T> implements Iterable<T> {

    T data;
    BinaryTree<T> left;
    BinaryTree<T> right;

    public BinaryTree() {
        data = null;
    }

    public BinaryTree(T t) {
        data = t;
        left = null;
        right = null;
    }

    public void addLeftChild(BinaryTree<T> t) {
        left = t;
    }

    public void addRightChild(BinaryTree<T> t) {
        right = t;
    }

    @Override
    public Iterator<T> iterator() {
        return new BinaryTreeIterator(this);
    }

    private class BinaryTreeIterator implements Iterator<T> {

        private LinkedList<BinaryTree<T>> list = new LinkedList<>();

        BinaryTreeIterator(BinaryTree<T> tree) {
            if (tree.data != null)
                list.add(tree);
        }

        @Override
        public boolean hasNext() {
            return !list.isEmpty();
        }

        @Override
        public T next() {
            if (!hasNext())
                throw new NoSuchElementException();
            BinaryTree<T> tree = list.remove();
            if (tree.left != null)
                list.add(tree.left);
            if (tree.right != null)
                list.add(tree.right);
            return tree.data;
        }
    }
}
